### PR TITLE
fix(cli): skip browser + loopback on cross-device auth login (SUPER-750)

### PR DESCRIPTION
**Pull request by justincrich**
## Summary

`superset auth login` opens a local browser for the OAuth loopback flow. On SSH / remote-workspace / no-DISPLAY contexts the local browser is on a different machine than the CLI, so the loopback redirect cannot reach the CLI's port — the user sees a doomed browser tab, then has to abort and fall back to `--api-key`.

This PR:
- Widens `shouldOpenBrowser()` to detect `SUPERSET_WORKSPACE_ID` and Linux headless (`!DISPLAY && !WAYLAND_DISPLAY`)
- Gates `bindLoopbackServer()` inside the browser-open branch so it's skipped entirely in cross-device contexts (also closes a narrow shared-host attack surface — see security note)
- Adds an explicit `--no-browser` CLI flag for any context our heuristics miss
- Updates paste-primary copy to show the authorization domain ("Sign in to app.superset.sh using the link below") instead of misleading "Browser didn't open?"

Refs [SUPER-750](https://linear.app/superset-sh/issue/SUPER-750/cli-auth-login-dont-open-a-browser-in-cross-device-ssh-cases-fall-back). Original report: Daniel Vega (Inversion Semiconductor) hit this running Codex on an EC2 box.

## Changes

**Production (3 files + 1 new helper, ~70 LOC):**
- `packages/cli/src/lib/auth.ts` — `shouldOpenBrowser()` predicate widen; `LoginCallbacks.noBrowser?: boolean`; `login()` restructure to gate loopback bind
- `packages/cli/src/commands/auth/login/command.ts` — `--no-browser` flag declared and threaded; copy branch on `pasteOnly`
- `packages/cli/src/commands/auth/login/LoginUI.tsx` — `pasteOnly?: boolean` prop on `LoginUIProps`; domain-aware copy
- `packages/cli/src/commands/auth/login/derivePasteOnly.ts` — pure helper extracted to make AC-10 truth-table testable

**Tests (1 file, +200 LOC):**
- `packages/cli/src/lib/auth.test.ts` — 16 tests covering every AC
  - `shouldOpenBrowser` detection: SSH / workspace / Linux-headless / TTY / CI / darwin-no-DISPLAY (real env manipulation)
  - `LoginCallbacks` interface type-contract for `noBrowser`
  - `LoginUI` Ink render via `ink-testing-library` — asserts copy branches AND domain anchor visible
  - `derivePasteOnly` truth table — both trigger conditions covered independently (AC-10)

## Test plan

- [x] `bun test packages/cli/src/lib/auth.test.ts` — 16 pass, 0 fail
- [x] `bun run typecheck` — clean across all 28 packages
- [x] `bun x biome check` on touched files — clean
- [ ] Manual smoke: `SUPERSET_WORKSPACE_ID=test superset auth login` — verify no browser opens, paste flow surfaces with domain copy
- [ ] Manual smoke: `superset auth login --no-browser` on local TTY — same
- [ ] Manual smoke: `superset auth login` on local TTY — verify browser still opens (no regression)

## Security note

The security review (PKCE/state surface, shared-host attack surface, phishing risk, public-client model) found **no regression** — in fact, skipping `bindLoopbackServer()` in cross-device contexts **closes** a narrow shared-host attack surface (loopback HTTP handler accepting requests where `state` could be extracted from stdout scraping). PKCE/state cryptographic invariants are intact: both flows generate `codeVerifier` / `codeChallenge` / `state` before any bind, both use `code_challenge_method=S256`, both validate state at the same gate, both exchange with the same `codeVerifier` held in CLI process memory only. The `--no-browser` flag's paste-back habit is contained by PKCE (attacker-supplied `code#state` cannot be redeemed without the CLI's `code_verifier`). To reduce the residual UX footgun, paste-primary copy now displays the authorization host as a visual anchor before the paste field.

## Merge-conflict note

MEDIUM (line-offset, not structural) overlap with `super-752-host-auth-002` worktree: that branch adds imports + removes `LoginResult` and `refreshAccessToken` from `auth.ts`. The `shouldOpenBrowser()` and `login()` regions modified by this PR are identical to main in the 752 worktree — auto-merge expected to succeed cleanly on rebase.

## Pre-existing lint

Repo-wide `bun run lint` exits 1 due to an `organizeImports` warning in `packages/ui/src/assets/icons/preset-icons/index.ts` (around `mastracode.svg` import). This is present on `main` and was on the base commit before this branch was created. Not introduced by this PR; flagged for a separate fix.

## Out of scope (filed as follow-ups)

- Typed `CrossDeviceReason` enum + reason-specific UI copy (architectural polish; separate refactor ticket if detection grows)
- Safari OAuth redirect failure Daniel also reported (separate investigation; root cause is not in this code — `buildAuthorizeUrl()` already builds `response_type=code` deterministically for both URLs)
- `openBrowser()` silent failure on Linux (`auth.ts:320` fire-and-forget `void`)

---

<a href="https://stagereview.app/superset-sh/superset/pull/4884">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

---
## Summary by cubic
Automatically skips opening a browser and binding the loopback server for `superset auth login` in SSH/headless/cross‑device contexts, using a paste‑only flow with clear, domain‑aware copy. Removes the `--no-browser` flag in favor of auto‑detection and reduces the attack surface. Addresses SUPER-750.

- **Bug Fixes**
  - Broaden `shouldOpenBrowser()` detection: keep SSH/CI/TTY checks, add Linux headless (`!DISPLAY` and `!WAYLAND_DISPLAY`), remove the `SUPERSET_WORKSPACE_ID` guard; export for reuse.
  - Gate `bindLoopbackServer()` behind the browser-open decision; skip entirely in paste‑only mode.
  - Show domain‑aware copy in Ink UI and the non‑Ink fallback; thread `pasteOnly` into `LoginUI`.

- **Refactors**
  - Remove `--no-browser` and the `derivePasteOnly` helper; use `!shouldOpenBrowser()` to drive paste‑only.
  - Expand tests for SSH/headless detection and `LoginUI` copy; restore the `refreshAccessToken` sanitization regression test; add `ink-testing-library` for component tests.

<sup>Written for commit 9510c203d202b5454476cdedb83904045e82ad46. Summary will update on new commits. <a href="https://cubic.dev/pr/superset-sh/superset/pull/4884?utm_source=github">Review in cubic</a></sup>

## Summary by CodeRabbit

* **New Features**
  * Added --noBrowser flag to skip opening a browser and force a paste-only sign-in flow.
  * Login prompts now show host-aware paste instructions when paste-only mode is used.
  * Improved fallback behavior: when a browser can’t be opened, the flow prefers the paste/code URL and clearer instructions.

* **Tests**
  * Expanded tests for login behavior and browser-availability detection across environments and paste-only scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/superset-sh/superset/pull/4884?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/cli/src/lib/auth.test.ts [hunk 1]
```diff
@@ -1,44 +1,224 @@
...
+		expect(result).toBe(false);
+	});
+
+	test("AC-2c: shouldOpenBrowser returns true when SSH_* not set (regression baseline)", () => {
+		process.env = {};
+		process.stdout.isTTY = true;
+
+		const result = shouldOpenBrowser();
+		expect(result).toBe(true);
+	});
```
[the remark this answers is not in the document]
Comment (justincrich) [reply packages/cli/src/lib/auth.test.ts:69]: Addressed in 4a8cb44b8. All three AC-2 SSH tests (AC-2a/b/c) now pin `process.platform` to `darwin` before exercising the env-var-only paths, so the SSH detection assertions are isolated from the headless-Linux branch and will run identically on macOS dev machines and Linux CI runners.

File: packages/cli/src/lib/auth.test.ts [hunk 2]
```diff
@@ -1,44 +1,224 @@
-import { afterEach, describe, expect, mock, test } from "bun:test";
-import { CLIError } from "@superset/cli-framework";
-import { refreshAccessToken } from "./auth";
+import { afterEach, beforeEach, describe, expect, test } from "bun:test";
+import { type LoginCallbacks, shouldOpenBrowser } from "./auth";
```
[the remark this answers is not in the document]
Comment (justincrich) [reply packages/cli/src/lib/auth.test.ts]: Addressed in a96d0478b. Restored the `sanitizes OAuth refresh failure details` test verbatim from cf3330e3e (SUPER-752 #4738) and re-imported `refreshAccessToken` + `CLIError`. The test now lives at the bottom of `auth.test.ts` and asserts that a 400 response containing `access_token`, `refresh_token`, `redirect` with auth code, and `cookie` values produces a `CLIError` whose `message + suggestion` contain none of those secrets. `bun test` is green (17/17).

File: packages/cli/src/lib/auth.test.ts [hunk 3]
```diff
@@ -1,44 +1,224 @@
...
+			SSH_TTY: "pts/0",
+		};
+		process.stdout.isTTY = true;
+
+		const result = shouldOpenBrowser();
+		expect(result).toBe(false);
+	});
+
+	test("AC-2c: shouldOpenBrowser returns true when SSH_* not set (regression baseline)", () => {
+		process.env = {};
```
[the remark this answers is not in the document]
Comment (justincrich) [reply packages/cli/src/lib/auth.test.ts:64]: Addressed in 4a8cb44b8 — same root cause as the greptile P1 above. `process.platform` is now pinned to `darwin` in AC-2a/b/c so SSH detection is isolated from the headless-Linux branch on CI.

File: .spec/improvements/SUPER-869/BRIEF.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+**BUILTIN_TERMINAL_AGENTS entry:**
+
+```
+id: "droid"
+label: "Droid"
+description: "Factory's agent-native coding CLI for autonomous software development."
+command: "droid --auto medium"
+promptCommand: "droid --auto medium"
+```
```
[the remark this answers is not in the document]
Comment (justincrich) [reply .spec/improvements/SUPER-869/BRIEF.md]: Skipping — `.spec/improvements/SUPER-869/BRIEF.md` was an orchestration artifact that should not have been in this PR's diff in the first place (the SUPER-869 Droid preset work is a separate effort unrelated to SUPER-750). The file was removed from the PR in 338e0f0cc (`chore(SUPER-750): remove orchestration artifacts before PR`), so the MD040 fence-language issue no longer applies here.

File: packages/cli/src/commands/auth/login/command.ts [hunk 5]
```diff
@@ -152,9 +153,14 @@ export default command({
 		apiKey: string().desc(
 			"Store a Superset API key (sk_live_…) at ~/.superset/config.json instead of running the OAuth flow",
 		),
+		noBrowser: boolean().desc(
```
Comment (saddlepaddle) [inline packages/cli/src/commands/auth/login/command.ts]: Hmm I think we actually don't need to expose this as part of the cli's api - wdyt? Mainly making that call bc a lot of the other competitors simply detect this correctly

File: packages/cli/src/lib/auth.ts [hunk 6]
```diff
@@ -55,10 +56,18 @@ export function getWebUrl(): string {
 	return env.SUPERSET_WEB_URL;
 }
 
-function shouldOpenBrowser(): boolean {
+export function shouldOpenBrowser(): boolean {
 	if (!process.stdout.isTTY) return false;
 	if (process.env.CI) return false;
 	if (process.env.SSH_CONNECTION || process.env.SSH_TTY) return false;
+	if (process.env.SUPERSET_WORKSPACE_ID) return false;
```
Comment (saddlepaddle) [inline packages/cli/src/lib/auth.ts]: Hmm and I guess this is okay bc if you're in a superset workspace, we've already injected a valid version of the superset cli?

File: packages/cli/src/commands/auth/login/command.ts (hunk 5 above)
Comment (justincrich) [reply packages/cli/src/commands/auth/login/command.ts]: Agreed — dropped `--no-browser` from the CLI surface. `shouldOpenBrowser()` already detects every case where the browser can't open (TTY, CI, SSH, headless-Linux), and the paste flow runs as the always-on fallback regardless.

Changes in 9510c203d:
- Removed `--no-browser` option from `auth login`
- Removed `noBrowser` from `LoginCallbacks`
- Deleted `derivePasteOnly` helper — `pasteOnly` is now just `!shouldOpenBrowser()`
- Dropped AC-7 and AC-10 tests

Net: -81 lines, fewer concepts, same behavior except in the corner case where the user wanted to force paste-only on a local TTY (which competitors don't support either).

Typecheck + lint + tests green.
